### ISSUE TITLE
[BUG] `plist` resource does not properly work with `dict` types

### ISSUE THREAD

**Issue by kquinsland** [microsoft/macos-cookbook#215]
## Describe the Bug 
I am unable to use the `plist` resource to create a `dict`.

There are *two* issues  here.  

1. the property `value` [does  *not* support `Hash`](https://github.com/microsoft/macos-cookbook/blob/2d5f5a365d3df597a7573c4d4dd6e7993dd5c5cf/resources/plist.rb#L5), even though [`type_to_commandline_string`](https://github.com/microsoft/macos-cookbook/blob/2d5f5a365d3df597a7573c4d4dd6e7993dd5c5cf/libraries/plist.rb#L31) does  support the `hash` type.

2. The `PlistBuddy` command that eventually does get  executed does  not have correct syntax for hash operations anyways.

here's a quick example to illustrate my point; i am trying to set  the first day of the week to  monday. This is done by setting the `AppleFirstWeekday` key in the `~/Library/Preferences/.GlobalPreferences.plist` file to a `dict` with the value `{ gregorian = 2;}`

As you can see here:
```
$ defaults read -g AppleFirstWeekday                                                                                          
{
    gregorian = 2;
}
```

The goal is to find an equivalent `plist` block that does the same thing.  Here's what i have now:

```
# Set First day of week to Monday
# defaults write -g AppleFirstWeekday -dict gregorian 2
##
plist 'make Monday the first DOW' do
    path '/Users/somebody/Library/Preferences/.GlobalPreferences.plist'
    entry 'AppleFirstWeekday'
    value "{ 'gregorian' => 2;}"
    owner 'somebody'
    group 'staff'
end
``` 

Which does not work. I get entries that look like this in the chef run:

```
      - execute /usr/libexec/PlistBuddy -c 'Set :"AppleFirstWeekday" { "gregorian" => 2}' "/Users/somebody/Library/Preferences/.GlobalPreferences.plist"
```

Which is incorrect. If i pass in a true ruby Hash, i get a validation error; see issue 1 above.

The *correct* syntax for `dicts` when using `PlistBuddy` is this:

```
$ /usr/libexec/PlistBuddy -c 'Set :"AppleFirstWeekday":gregorian 4' "/Users/somebody/Library/Preferences/.GlobalPreferences.plist"
$ defaults read -g AppleFirstWeekday                                                                                          
{
    gregorian = 4;
}
```

## Error Message  
There is no error, only incorrect behavior.

## Chef Exception  
There is no exception, only incorrect behavior.

**Comment by americanhanko** [comment]
@kquinsland thanks for the bug. Fixed in b89bfd25e499608bedfa7fd2bb74e25c18a2120b (at least the unit tests say so 🙂). YSK that support for this is pretty basic; nested dicts or arrays won't work because PlistBuddy has some odd syntax for that (e.g. `KeyFoo:0:KeyBar:1`). This should get merged into the next release branch, though.

**Comment by kquinsland** [comment]
@americanhanko thanks for the quick turn around! 


> YSK that support for this is pretty basic; nested dicts or arrays won't work because PlistBuddy has some odd syntax for that 

Yep! It took me a few min to figure out how the `plist` was implemented / what it was doing under the hood to figure out the *correct* syntax for plist buddy...but after i figured out which binaries  you were calling i was able to reverse the correct syntax for plist buddy and compare it to "known working" input/output from  `defaults`.

As a workaround, i just use an `execute 'defaults write "#{blah}' ` block for the payloads that are reasonably complex

> This should get merged into the next release branch, though.

So happy to hear that. Looking forward to getting some of the `execute` blocks removed :)
